sagemathinc/cocalc
# webapp: show OOM events to user and some fixes related to session restore

There is an entry in the DB showing OOM kills of the linux kernel's quota management. This patch exposes that information to the user. I know that the way how this is recorded will change, but right now the DB is still updated.

The patch itself shows how many such oom events did happen, and after being dismissed and additional ones are happening, a (higher-priority) alert is presented to the user.

Incidentally, the necessity to reset some state value of a project upon restarting it, required a fix of the entire procedure how sessions are restored and file-targets are opened up. After that, I converted session to typescript, too.

This also includes https://github.com/sagemathinc/cocalc/pull/3429 (which is pulled out to be reviewed separately)

Before making this live, the kucalc manager needs to be updated, too: https://github.com/sagemathinc/kucalc/commit/9674902acf94acbfba044119f29f4b2d3304760d

## Testing
* session related
  * delete all local storage `session./<test project id` and `closed-session...` in the debugger
  * then open up a specific file path, refresh the page and check the session was stored correctly
  * close a project and open the same project again (tabs open up based on`closed-session`)
  * there is also no random session set for fullscreen popups, i.e. shift-click on an editor tab and check that there is just one editor in one project
  * when restoring a session after refreshing the page, previously opened and closed tabs in `closed-session` are no longer restored ... just the session as it is
  * the current editor tab is focused after restart, and the ordering of project-tabs is preserved.

* OOM related. the DB lines below simulate OOM events by editing the `oom_kills` value in the project `status`. The dismissed-value also includes the `start_ts` value to be matched against the current project session.
  * first time: a warning, dismiss it
  * increase the oom kills value, tells user how many more happend
  * increase `start_ts` to simulate a project restart

in the DB:

```
-- set a start_ts
UPDATE projects SET status = jsonb_set(status, '{start_ts}', '1547438878195'::JSONB) WHERE project_id='[ test project UUID ]';
UPDATE projects SET status = jsonb_set(status, '{oom_kills}', '1'::JSONB) WHERE project_id='[ test project UUID ]';
-- dismiss it, and 2 more
UPDATE projects SET status = jsonb_set(status, '{oom_kills}', '3'::JSONB) WHERE project_id='[ test project UUID ]';
-- increase start_ts slightly (new project session)
UPDATE projects SET status = jsonb_set(status, '{start_ts}', '1547458878195'::JSONB) WHERE project_id='[ test project UUID ]';
```

* **testing on K2**: regarding actual testing, I deployed the change to "manager" and webapp on K2. I used a simple Python3 notebook with that code to trigger it under real-word cirumstances:
```
import numpy as np
arrays = []
```
and then adding large arrays. the first one still works, the second line triggers the OOM.
```
arrays.append(np.random.randn(120000000))
arrays.append(np.random.randn(120000000))
```
